src/Microsoft.VisualStudio.Web.CodeGeneration.MsBuild/MsBuilder.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+
+namespace Microsoft.VisualStudio.Web.CodeGeneration.MsBuild
+{
+    public class MsBuilder<T> where T : IMsBuildProcessor
+    {
+        private readonly Project _project;
+        private T _buildProcessor;
+
+        public MsBuilder(string filePath, T buildProcessor)
+        {
+            if (buildProcessor == null)
+            {
+                throw new ArgumentNullException(nameof(buildProcessor));
+            }
+            _buildProcessor = buildProcessor;
+            _buildProcessor.Init();
+            _project = CreateProject(filePath, _buildProcessor.Configuration);
+        }
+
+        public void RunMsBuild()
+        {
+            var result = RunMsBuild(_project);
+            _buildProcessor.ProcessBuildResult(_project, result.ProjectStateAfterBuild, result.ResultsByTarget);
+        }
+
+        private T BuildProcessor
+        {
+            get
+            {
+                return _buildProcessor;
+            }
+        }
+
+        private BuildResult RunMsBuild(Project project)
+        {
+            var projectInstance = project.CreateProjectInstance();
+            //EnsureTargetDirectoryExists(projectInstance);
+
+            var targets = BuildProcessor.Targets;
+
+            var buildRequest = new BuildRequestData(projectInstance, targets, null, BuildRequestDataFlags.ProvideProjectStateAfterBuild);
+            var buildParams = new BuildParameters(project.ProjectCollection);
+
+            // Add loggers
+            buildParams.Loggers = BuildProcessor.Loggers;
+            var result = BuildManager.DefaultBuildManager.Build(buildParams, buildRequest);
+
+            if (result.ProjectStateAfterBuild == null)
+            {
+                // this is a hack for failed project builds. ProjectStateAfterBuild == null after a failed build
+                // But the properties are still available to be read
+                result.ProjectStateAfterBuild = projectInstance;
+            }
+
+            return result;
+        }
+
+        private Project CreateProject(string filePath, string configuration)
+        {
+            var sdkPath = new DotNetSdkResolver().ResolveLatest();
+            var msBuildFile = RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
```

Comment:
It turns out that this need to be "MSBuild.exe" even on Linux. (https://github.com/natemcmaster/ProjectModelAbstractions/commit/4fe4021031d508662ede99377bfde5d00af37db2)